### PR TITLE
RES-2418: break dependencies by project

### PR DESCRIPTION
**Pull request by lscheinkman**
This PR separates `nupic.research` dependencies into main framework dependencies and project dependencies. Now each project maintains its own dependencies as [optional dependencies](https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#optional-dependencies) that can be installed independently of the other project's dependencies. To install project specific dependencies, `pip install` must be called with the `extra` modifier specifying the optional dependencies, for example:
```
pip install -e ".[dendrites]"
```
or 
```
pip install -e ".[transformers]"
```

The main `conda` environment install all `vernon` based projects (except `transformers`) so it should be backward compatible with the current environments.

### CODE REVIEW

File: projects/visual_recognition_grid_cells/python2_htm_docker/docker_dir/train_and_eval_GridCellNet.py [hunk 1]
```diff
@@ -73,21 +73,21 @@
...
-    columnCount,
-    featuresPerObject,
-    objectWidth,
-    numModules,
-    seed1,
-    seed2,
-    anchoringMethod):
+        EVAL_ON_TRAINING_DATA_BOOL,  # noqa: N803
+        locationModuleWidth,
+        feature_columns_to_grid_cells_threshold,
```
Comment (lucasosouza) [inline projects/visual_recognition_grid_cells/python2_htm_docker/docker_dir/train_and_eval_GridCellNet.py:78]: If the `inferObjectWithRandomMovements` had noqa: N803 for all camel-cased attributes, shouldn't it be the same here? Or have it on a file level?

File: environment.yml [hunk 2]
```diff
@@ -4,26 +4,28 @@
...
+
+      # Install nupic.research in development mode with dev tools extras enabled
+      - -e ".[dev]"
 
-      # Install nupic.research in develop mode
-      - -r requirements-dev.txt
-      - -e . --no-binary :all
+      # Install projects extra dependency
+      # See `setup.cfg` for available extras.
+      - -e ".[vernon,imagenet,gsc,dendrites]"
```
Comment (lucasosouza) [inline environment.yml:28]: Is the implication that imagenet and gsc are installed by default when installing nupic.research? Should we make this assumption? 

I would argue moving forward we are working a lot more with dendrites, transformers and embodied. imagenet and gsc are more legacy projects.

**Review by lucasosouza — APPROVED** [review]
Awesome, this should help a lot. I just had a comment regarding pep8 exception and a question regarding what should be the basic environments loaded when installing nupic.research

File: environment.yml (hunk 2 above)
Comment (lscheinkman) [reply environment.yml:28]: When you install `nupic.research` without any projects, only the dependencies listed in the [install_requires](https://github.com/numenta/nupic.research/pull/569/files#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52R50-R61) section will be installed by default

This file is only relevant when creating a new `conda` environment not the `nupic.research` package itself. 

I am installing the current projects to keep the current `conda` environment compatible with this PR. But I agree with you in the future we should deprecate these projects and remove them from the `conda` environment as well.

Comment (lscheinkman) [reply environment.yml:28]: Another option, is to keep the default `conda` environment clean from any project dependency and let the end user install the projects they use on-demand.

File: projects/visual_recognition_grid_cells/python2_htm_docker/docker_dir/train_and_eval_GridCellNet.py (hunk 1 above)
Comment (lscheinkman) [reply projects/visual_recognition_grid_cells/python2_htm_docker/docker_dir/train_and_eval_GridCellNet.py:78]: I'm not really sure how this file passed `flake8` and code review.
IMHO, we should not mix snake-case and camel-case.

**Pull request merged**